pkg/services/ngalert/state/cache.go

```diff
@@ -19,7 +19,8 @@ import (
 )
 
 type ruleStates struct {
-	states map[string]*State
+	ruleVersion int64
```

Comment:
I know we talked about changing the state cache to index by version, but I'm not sure I understand why that change is necessary for this fix?

Reply:
This is the last part of the solution. Please refer to PR description